File: BinaryTrees/binary_trees.py
# binary_trees.py
"""Volume 2: Binary Trees.
Caelan Osman
Math 321 sec 3
October 8, 2020
"""

# These imports are used in BST.draw().
import networkx as nx
from networkx.drawing.nx_agraph import graphviz_layout
from matplotlib import pyplot as plt
import random
import time


class SinglyLinkedListNode:
    """A node with a value and a reference to the next node."""
    def __init__(self, data):
        self.value, self.next = data, None

class SinglyLinkedList:
    """A singly linked list with a head and a tail."""
    def __init__(self):
        self.head, self.tail = None, None

    def append(self, data):
        """Add a node containing the data to the end of the list."""
        n = SinglyLinkedListNode(data)
        if self.head is None:
            self.head, self.tail = n, n
        else:
            self.tail.next = n
            self.tail = n

    def iterative_find(self, data):
        """Search iteratively for a node containing the data.
        If there is no such node in the list, including if the list is empty,
        raise a ValueError.

        Returns:
            (SinglyLinkedListNode): the node containing the data.
        """
        current = self.head
        while current is not None:
            if current.value == data:
                return current
            current = current.next
        raise ValueError(str(data) + " is not in the list")

    # Problem 1
    def recursive_find(self, data):
        """This function searches recursively for the node containing the data.
        Returns:
            (SinglyLinkedListNode): the node containing the data.
        raises:
            (ValueError): If the list is empty or the data is not contained in the list. 
        """
        #function to revursively find node
        def is_node(node):
            #if the SLL is empty raise a value error
            if node == None:
                raise ValueError('The list is empty')
            #if current node contains the data return that node
            elif node.value == data:
                return node
            #if the next node is none then we know we are at the end of our list and no other node
            #contained the value so we can rais a value error that the data isn't contained in the SLL
            elif node.next == None:
                raise ValueError('The data is not contained in the list')
            #recursive call on the next node in the lsit
            else:
                return is_node(node.next)

        #initial call on the head node in the SLL
        return is_node(self.head)

class BSTNode:
    """A node class for binary search trees. Contains a value, a
    reference to the parent node, and references to two child nodes.
    """
    def __init__(self, data):
        """Construct a new node and set the value attribute. The other
        attributes will be set when the node is added to a tree.
        """
        self.value = data
        self.prev = None        # A reference to this node's parent node.
        self.left = None        # self.left.value < self.value
        self.right = None       # self.value < self.right.value


class BST:
    """Binary search tree data structure class.
    The root attribute references the first node in the tree.
    """
    def __init__(self):
        """Initialize the root attribute."""
        self.root = None

    def find(self, data):
        """Return the node containing the data. If there is no such node
        in the tree, including if the tree is empty, raise a ValueError.
        """

        # Define a recursive function to traverse the tree.
        def _step(current):
            """Recursively step through the tree until the node containing
            the data is found. If there is no such node, raise a Value Error.
            """
            if current is None:                     # Base case 1: dead end.
                raise ValueError(str(data) + " is not in the tree.")
            if data == current.value:               # Base case 2: data found!
                return current
            if data < current.value:                # Recursively search left.
                return _step(current.left)
            else:                                   # Recursively search right.
                return _step(current.right)

        # Start the recursion on the root of the tree.
        return _step(self.root)

    # Problem 2
    def insert(self, data):
        """Insert a new node containing the specified data.

        Raises:
            ValueError: if the data is already in the tree.
        """
        #recurssive function to step through BST and insert node where we want it to be
        def _step(node):
            if data < node.value: #if the data is less than the node value we insert to the left
                if node.left == None: #if there is no left node insert the data here
                    new = BSTNode(data)
                    node.left = new
                    new.prev = node
                else: #otherwise recursive call to _step(node) function
                    _step(node.left)

            #if the data is grater than the node value do the same thing about except on the right.
            #notice node.value != data because of the try except block below
            else:
                if node.right == None:
                    new = BSTNode(data)
                    node.right = new
                    new.prev = node
                else:
                    _step(node.right)

        try: #try to find the data
            self.find(data)
        except ValueError: #if the data is not found insert our new node
            if self.root == None: #if there is no root, the BST is empty, make the new node the root
                self.root = BSTNode(data)
            else: #otherwise call the step function
                _step(self.root)
        else: #if the data is found raise a value error about duplicates
            raise ValueError('Data already contained in tree, no duplicates allowed')

    # Problem 3
    def remove(self, data):
        """This function removes the node containing the specified data.
        Raises:
            ValueError: if there is no node containing the data, including if
                the tree is empty.
        """
        #if the BST is empty raise value error
        if self.root == None:
           raise ValueError('The BST is empty.')

        #this is the node to remove; self.find(data) will raise a value error
        #if there is no node in the BST containing the data
        node_to_remove = self.find(data)

        if node_to_remove.left == None and node_to_remove.right == None: #if the node to remove is a leaf node
            if node_to_remove.prev == None: #if it is the root node
                self.root = None
            elif node_to_remove.prev.left == node_to_remove: #if it is a left child of the parent node set the left pointer to None
                node_to_remove.prev.left = None
            else: #otherwise it is a right child so set the parents right child to none
                node_to_remove.prev.right = None
        elif node_to_remove.right == None: #if the node to remove has only a left child
            if self.root == node_to_remove: #if this node happens to be the root
                self.root = self.root.left
                self.root.prev = None
            elif node_to_remove.prev.left == node_to_remove: #if the node to remove is a left child of its parent node
                prev_node = node_to_remove.prev
                prev_node.left = node_to_remove.left
                prev_node.left.prev = prev_node
            else: #otherwise it is a right child of the parent node
                prev_node = node_to_remove.prev
                prev_node.right = node_to_remove.left
                prev_node.right.prev = prev_node
        elif node_to_remove.left == None: #if the node to remove has only a right child
            if self.root == node_to_remove: #if this node happens to be the root
                self.root = self.root.right
                self.root.prev = None
            elif node_to_remove.prev.left == node_to_remove: #if the node to remove is a left child of its parent node
                prev_node = node_to_remove.prev
                prev_node.left = node_to_remove.right
                prev_node.left.prev = prev_node
            else: #otherwise it is a right child of the parent node
                prev_node = node_to_remove.prev
                prev_node.right = node_to_remove.right
                prev_node.right.prev = prev_node
        else: #If the node to remove has two children or is the root node
            if self.root == node_to_remove: #if the node to remove is the root node (and has two children)
                leaf = self.root.left
                while(leaf.right != None): #go to the farthest right leaf node after going one level left
                    leaf = leaf.right
                new_value = leaf.value
                self.remove(leaf.value) #remove the leaf
                self.root.value = new_value #set the roots new value
            else: #if the node to remove is not the root, follow the same strategy as above
                leaf = node_to_remove.left
                while (leaf.right != None):
                    leaf = leaf.right
                new_value = leaf.value
                self.remove(leaf.value)
                node_to_remove.value = new_value


    def __str__(self):
        """String representation: a hierarchical view of the BST.

        Example:  (3)
                  / \     '[3]          The nodes of the BST are printed
                (2) (5)    [2, 5]       by depth levels. Edges and empty
                /   / \    [1, 4, 6]'   nodes are not printed.
              (1) (4) (6)
        """
        if self.root is None:                       # Empty tree
            return "[]"
        out, current_level = [], [self.root]        # Nonempty tree
        while current_level:
            next_level, values = [], []
            for node in current_level:
                values.append(node.value)
                for child in [node.left, node.right]:
                    if child is not None:
                        next_level.append(child)
            out.append(values)
            current_level = next_level
        return "\n".join([str(x) for x in out])

    def draw(self):
        """Use NetworkX and Matplotlib to visualize the tree."""
        if self.root is None:
            return

        # Build the directed graph.
        G = nx.DiGraph()
        G.add_node(self.root.value)
        nodes = [self.root]
        while nodes:
            current = nodes.pop(0)
            for child in [current.left, current.right]:
                if child is not None:
                    G.add_edge(current.value, child.value)
                    nodes.append(child)

        # Plot the graph. This requires graphviz_layout (pygraphviz).
        nx.draw(G, pos=graphviz_layout(G, prog="dot"), arrows=True,
                with_labels=True, node_color="C1", font_size=8)
        plt.show()

class AVL(BST):
    """Adelson-Velsky Landis binary search tree data structure class.
    Rebalances after insertion when needed.
    """
    def insert(self, data):
        """Insert a node containing the data into the tree, then rebalance."""
        BST.insert(self, data)      # Insert the data like usual.
        n = self.find(data)
        while n:                    # Rebalance from the bottom up.
            n = self._rebalance(n).prev

    def remove(*args, **kwargs):
        """Disable remove() to keep the tree in balance."""
        raise NotImplementedError("remove() is disabled for this class")

    def _rebalance(self,n):
        """Rebalance the subtree starting at the specified node."""
        balance = AVL._balance_factor(n)
        if balance == -2:                                   # Left heavy
            if AVL._height(n.left.left) > AVL._height(n.left.right):
                n = self._rotate_left_left(n)                   # Left Left
            else:
                n = self._rotate_left_right(n)                  # Left Right
        elif balance == 2:                                  # Right heavy
            if AVL._height(n.right.right) > AVL._height(n.right.left):
                n = self._rotate_right_right(n)                 # Right Right
            else:
                n = self._rotate_right_left(n)                  # Right Left
        return n

    @staticmethod
    def _height(current):
        """Calculate the height of a given node by descending recursively until
        there are no further child nodes. Return the number of children in the
        longest chain down.
                                    node | height
        Example:  (c)                  a | 0
                  / \                  b | 1
                (b) (f)                c | 3
                /   / \                d | 1
              (a) (d) (g)              e | 0
                    \                  f | 2
                    (e)                g | 0
        """
        if current is None:     # Base case: the end of a branch.
            return -1           # Otherwise, descend down both branches.
        return 1 + max(AVL._height(current.right), AVL._height(current.left))

    @staticmethod
    def _balance_factor(n):
        return AVL._height(n.right) - AVL._height(n.left)

    def _rotate_left_left(self, n):
        temp = n.left
        n.left = temp.right
        if temp.right:
            temp.right.prev = n
        temp.right = n
        temp.prev = n.prev
        n.prev = temp
        if temp.prev:
            if temp.prev.value > temp.value:
                temp.prev.left = temp
            else:
                temp.prev.right = temp
        if n is self.root:
            self.root = temp
        return temp

    def _rotate_right_right(self, n):
        temp = n.right
        n.right = temp.left
        if temp.left:
            temp.left.prev = n
        temp.left = n
        temp.prev = n.prev
        n.prev = temp
        if temp.prev:
            if temp.prev.value > temp.value:
                temp.prev.left = temp
            else:
                temp.prev.right = temp
        if n is self.root:
            self.root = temp
        return temp

    def _rotate_left_right(self, n):
        temp1 = n.left
        temp2 = temp1.right
        temp1.right = temp2.left
        if temp2.left:
            temp2.left.prev = temp1
        temp2.prev = n
        temp2.left = temp1
        temp1.prev = temp2
        n.left = temp2
        return self._rotate_left_left(n)

    def _rotate_right_left(self, n):
        temp1 = n.right
        temp2 = temp1.left
        temp1.left = temp2.right
        if temp2.right:
            temp2.right.prev = temp1
        temp2.prev = n
        temp2.right = temp1
        temp1.prev = temp2
        n.right = temp2
        return self._rotate_right_right(n)


# Problem 4
def prob4():
    """This function compares the build and search times of the SLL, BST, and AVL classes
    """

    infile = 'english.txt'
    with open(infile, 'r') as ip: #add lines to list
        data = ip.readlines()
    num_elements = [2**n for n in range(3, 11)] #number of random elements
    #sll times
    sll_load_times = []
    sll_find_times = []
    #bst times
    bst_load_times = []
    bst_find_times = []
    #avl times
    avl_load_times = []
    avl_find_times = []

    for n in num_elements: #computing times
        random_elements = random.sample(data, n) #List containing n random elements from data
        subset = random.sample(random_elements, 5) #subset containing 5 random elements from random_elements


        #timing creation times
        #sll
        sll = SinglyLinkedList()
        sll_start = time.time()
        for element in random_elements:
            sll.append(element)
        sll_end = time.time()
        sll_load_times.append(sll_end - sll_start)

        #bst
        bst = BST()
        bst_start = time.time()
        for element in random_elements:
            bst.insert(element)
        bst_end = time.time()
        bst_load_times.append(bst_end - bst_start)

        #avl
        avl = AVL()
        avl_start = time.time()
        for element in random_elements:
            avl.insert(element)
        avl_end = time.time()
        avl_load_times.append(avl_end - avl_start)



        #timing finding times
        #sll
        sll_start = time.time()
        for element in subset:
            sll.iterative_find(element)
        sll_end = time.time()
        sll_find_times.append(sll_end - sll_start)

        #bst
        bst_start = time.time()
        for element in subset:
            bst.find(element)
        bst_end = time.time()
        bst_find_times.append(bst_end - bst_start)

        #avl
        avl_start = time.time()
        for element in subset:
            avl.find(element)
        avl_end = time.time()
        avl_find_times.append(avl_end - avl_start)

    #plotting
    ax1 = plt.subplot(121)
    ax1.loglog(num_elements, sll_load_times, basex=2, basey=2, label='SLL Load Times')
    ax1.loglog(num_elements, bst_load_times, basex=2, basey=2, label='BST Load Times')
    ax1.loglog(num_elements, avl_load_times, basex=2, basey=2, label='AVL Load Times')
    ax1.legend(loc='upper left')
    plt.title('Load Times')
    plt.xlabel('Number of Elements in Structure')
    plt.ylabel('Toad Times')
    ax2 = plt.subplot(122)
    ax2.plot(num_elements, sll_find_times, label='SLL Find Times')
    ax2.plot(num_elements, bst_find_times, label='BST Find Times')
    ax2.plot(num_elements, avl_find_times, label='AVL Find Times')
    ax2.legend(loc='upper left')
    plt.title('Find Times')
    plt.xlabel('Number of Elements in Structure')
    plt.ylabel('Find Times')
    plt.show()

File: Julia/BinaryTrees/binary_trees.jl
# binary_trees.jl

module BinaryTrees

import Base.append!, Base.insert!

using Plots, Graphs, GraphRecipes

export SinglyLinkedListNode, SinglyLinkedList, append!, iterativeFind,recursiveFind, BSTNode, BST, AVL, find, insert!, remove!, draw

mutable struct SinglyLinkedListNode
    """
    A node with a value and a reference to the next node
    """
    data::Union{<:Number, String, Vector{<:Any}}
    next::Union{Nothing,SinglyLinkedListNode}

    function SinglyLinkedListNode(data=nothing, next=nothing)
        """
        constructor if no data and no next node is passed in. I.e. SinglyLinkedListNode()
        """
        return new(data, next)
    end
    function SinglyLinkedListNode(data; next=nothing)
        """
        constructor if no next node is passed in. I.e. SinglyLinkedListNode([3])
        """
        return new(data, next)
    end
    function SinglyLinkedListNode(data,next)
        """
        Constructor if all information is passed in I.e. SinglyLinkedListNode([3], nextNode)
        """
        return new(data,next)
    end
end

mutable struct SinglyLinkedList
    """
    A singly linekd list with a head and a tail
    """
    head::Union{Nothing, SinglyLinkedListNode}
    tail::Union{Nothing, SinglyLinkedListNode}
    function SinglyLinkedList(;head=nothing, tail=nothing)
        """
        Constructor if no data is passed in i.e. SinglyLinkedList()
        """
        return new(head,tail)
    end
    function SinglyLinkedList(head;tail=nothing)
        """
        Constructor if only the head is passed in.
        """
        return new(head, head)
    end
    function SinglyLinkedList(head, tail)
        """
        Constructor if all the data is passed in: i.e. SinglyLinkedList(head, tail)
        """
        if isequal(head,nothing) && isequal(tail,nothing)
            return new(nothing, nothing)
        elseif isequal(head.next, nothing)
            throw(ArgumentError("The head is not connected to the tail"))
        end
        return new(head,tail)
    end

end

function append!(SLL::SinglyLinkedList, data)
    """
    Add a node containing th data to the end of the linked list.
    """
    # create the node
    newNode = SinglyLinkedListNode(data)

    if isequal(SLL.head,nothing) # if the linked list is empty.
        SLL.head=newNode
        SLL.tail=newNode
    else # otherwise add the node to the end of the list.
        SLL.tail.next = newNode
        SLL.tail = newNode
    end

    return
end

function iterativeFind(SLL::SinglyLinkedList, data)
    """
    Search iteratively for a node containing the data.
    If there is no such node in the list, including if the list is empty,
    throw a key error.
    """

    currentNode = SLL.head # start with the head node
    while !isequal(currentNode, nothing)
        # check if the data is in the current node and return the node if it is
        if currentNode.data == data
            return currentNode
        end
        # otherwise go to the next node
        currentNode = currentNode.next
    end

    # throw a key error if it is not found
    throw(KeyError(data))

    return
end

function recursiveFind(SLL::SinglyLinkedList, data)
    """
    We search recursively for the node containing the data and return the node.
    If there is no such node found or the list is empty a key error is thrown.
    """

    function isNode(currentNode::SinglyLinkedListNode)
        """
        This is the function used to recursively find the node
        """

        # if the current node is nothing then the list is empty (because of function construction -2nd elseif-
        # this if will only be activated on the head node)
        if isequal(currentNode,nothing)
            throw(KeyError(data))
        # check if the current node contains the data and return it
        elseif currentNode.data == data
            return currentNode
        # if the next node is nothing that means the data can't be contained in the list
        elseif isequal(currentNode.next,nothing)
            throw(KeyError(data))
        else # if none of the above happen, call this function on the next node
            return isNode(currentNode.next)
        end
    end
    # call function on head node.
    return isNode(SLL.head)
end

mutable struct BSTNode
    """
    A node class for binary search and avl trees. Contains a value,
    A reference to the parent node, and references to two child nodes.
    """
    data::Union{Nothing, <:Number, String, Vector{<:Any}}
    prev::Union{Nothing, BSTNode}
    left::Union{Nothing, BSTNode}
    right::Union{Nothing, BSTNode}

    function BSTNode(;data=nothing, prev=nothing,left=nothing,right=nothing)
        """
        Constructor if no data is provided. I.e. BSTNode()
        """
        return new(data,prev,left,right)
    end
    function BSTNode(data; prev=nothing,left=nothing,right=nothing)
        """
        Constructor if only the data is provided. I.e. BSTNode(data)
        """
        return new(data,prev,left,right)
    end
    function BSTNode(data,prev,left,right)
        """
        Constructor if all the data is provided. i.e. BSTNode(data,prev,left,right)
        """
        return new(data,prev,left,right)
    end
end

#=
TODO: Consider the following structure as a
replacement for the current architecture:

abstract type Tree end

mutable struct TreeStructureTest <: Tree
    root::Union{Nothing, BSTNode}
    nodeCount::Int64
end

mutable struct BST{T::TreeStructureTest}
    structure::TreeStructureTest
end

mutable struct AVL{T::TreeStructureTest}
    structure::TreeStructureTest

mutable struct FullTree{T<:TreeStructureTest}
    structure::T
    help::String
end
=#


# abstract type to build the BST and AVL structs off of
abstract type TreeStructure end

mutable struct BST <: TreeStructure
    """
    binary search tree data structure class.
    The root attribute references the first node in the tree.
    The node count is the length of the tree.
    """
    root::Union{Nothing, BSTNode}
    nodeCount::Int64

    function BST(;root=nothing,nodeCount=nothing)
        """
        Constructor if no data is provide. i.e. BST()
        """
        return new(root, 0)
    end

    function BST(root; nodeCount=0)
        """
            Constructor if the root is provided. i.e. BST(root).
        """
        if isequal(root,nothing)
            nodeCount = 0
        end
        return new(root,nodeCount)
    end

    function BST(root,nodeCount)
        """
        Constructor if the root and the node count is provided. i.e. BST(root, 3)
        Can be helpful if root already is linked to it's children.
        """
        if isequal(root,nothing)
            nodeCount = 0
        end
        return new(root,nodeCount)
    end
end

mutable struct AVL <: TreeStructure
    root::Union{Nothing, BSTNode}
    nodeCount::Int64

    function AVL(;root=nothing,nodeCount=nothing)
        """
        Constructor if no data is provide. i.e. AVL()
        """
        return new(root, 0)
    end
    function AVL(root; nodeCount=0)
        """
            Constructor if the root is provided. i.e. AVL(root).
        """
        if isequal(root,nothing)
            nodeCount = 0
        end
        newAVL = new(root,nodeCount)
        return _rebalance!(newAVL, newAVL.root)
    end

    function AVL(root,nodeCount)
        """
        Constructor if the root and the node count is provided. i.e. AVL(root, 3)
        Can be helpful if root already is linked to it's children.
        """
        if isequal(root,nothing)
            nodeCount = 0
        end

        return new(root,nodeCount)
    end
end


function find(B::TreeStructure, data)
    """
        Return the node containing the data. If there is no such node in the tree,
        including if the tree is empty, raise a KeyError.
    """

    # define recursive function to traverse the tree.
    function _step(currentNode)
        """ Recursively step through the tree until the node containing
            the data is found. If there is no such node, throw a KeyError.
        """
        if isequal(currentNode,nothing)
            throw(KeyError(data))
        elseif currentNode.data == data
            return currentNode
        elseif currentNode.data > data
            return _step(currentNode.left)
        else
            return _step(currentNode.right)
        end
    end
    return _step(B.root)
end

function _rebalance!(A::AVL, currentNode::Union{BSTNode, Nothing})
    # rebalance the subtree starting at the specified node

    balance = _balanceFactor(currentNode)

    if balance == -2
        if _height(currentNode.left.left) > _height(currentNode.left.right)
            currentNode = _rotate_left_left!(A, currentNode)
        else
            currentNode = _rotate_left_right!(A, currentNode)
        end
    elseif balance == 2
        if _height(currentNode.right.right) > _height(currentNode.right.left)
            currentNode = _rotate_right_right!(A, currentNode)
        else
            currentNode = _rotate_right_left!(A, currentNode)
        end
    end
    return currentNode
end


function _height(currentNode::Union{BSTNode, Nothing})

    if isequal(currentNode, nothing)
        return -1
    end

    return 1 + maximum((_height(currentNode.right), _height(currentNode.left)))
end

function _balanceFactor(currentNode::BSTNode)
    return _height(currentNode.right) - _height(currentNode.left)
end

function _rotate_left_left!(A::AVL, currentNode::Union{BSTNode, Nothing})
    temp = currentNode.left
    currentNode.left = temp.right
    if !isequal(temp.right, nothing)
        temp.right.prev = currentNode
    end

    temp.right = currentNode
    temp.prev = currentNode.prev
    currentNode.prev = temp

    if !isequal(temp.prev,nothing)
        if temp.prev.data > temp.data
            temp.prev.left = temp
        else
            temp.prev.right = temp
        end
    end

    if isequal(currentNode, A.root)
        A.root = temp
    end
    return temp
end

function _rotate_right_right!(A::AVL, currentNode::Union{BSTNode, Nothing})
    temp = currentNode.right
    currentNode.right = temp.left

    if !isequal(temp.left, nothing)
        temp.left.prev = currentNode
    end
    temp.left = currentNode
    temp.prev = currentNode.prev
    currentNode.prev = temp

    if !isequal(temp.prev,nothing)
        if temp.prev.data > temp.data
            temp.prev.left = temp
        else
            temp.prev.right = temp
        end
    end

    if isequal(currentNode,A.root)
        A.root = temp
    end

    return temp
end

function _rotate_left_right!(A::AVL, currentNode::Union{BSTNode, Nothing})
    temp1 = currentNode.left
    temp2 = temp1.right
    temp1.right = temp2.left

    if !isequal(temp2.left,nothing)
        temp2.left.prev = temp1
    end

    temp2.prev = currentNode
    temp2.left = temp1
    temp1.prev = temp2
    currentNode.left = temp2
    return _rotate_left_left!(A, currentNode)
end

function _rotate_right_left!(A::AVL, currentNode::Union{BSTNode, Nothing})
    temp1 = currentNode.right
    temp2 = temp1.left
    temp1.left = temp2.right

    if !isequal(temp2.right,nothing)
        temp2.right.prev = temp1
    end
    temp2.prev =currentNode
    temp2.right = temp1
    temp1.prev = temp2
    currentNode.right = temp2
    return _rotate_right_right!(A,currentNode)
end


function insert!(B::TreeStructure,data)
    """ Insert a new node containing the specified data.

        throws:
            KeyError: if the data is already in the tree.
    """

    function _step(currentNode)
        if data < currentNode.data # if the data is less than the node value we insert to the left
            if isequal(currentNode.left,nothing) # if there is no left node we insert the data here
                newNode = BSTNode(data)
                currentNode.left = newNode
                newNode.prev = currentNode
            else #otherwise recursive call to this function
                _step(currentNode.left)
            end
        else # if the data is greater than the node value do the same thing above except on the right
             # notice no duplicate values are allowed in the BST.
            if isequal(currentNode.right,nothing)
                newNode = BSTNode(data)
                currentNode.right = newNode
                newNode.prev = currentNode
            else
                _step(currentNode.right)
            end
        end
    end

    try # try and find the data
        find(B,data)
    catch e
        if isa(e, KeyError) # if the data is not found insert our new node
            if isequal(B.root, nothing) # if there is no root, the BST is empty, make the newnode the root
                B.root = BSTNode(data)
                B.nodeCount += 1
            else # otherwise call the step function to insert the node
                _step(B.root)
                B.nodeCount += 1
            end

            if isequal(typeof(B), AVL)
                node = find(B ,data)
                while !isequal(node,nothing)
                    node = _rebalance!(B,node).prev
                end
            end

        else # if the wrong error is caught, throw the error again.
            throw(e)
        end
    else
        # if no error is caught at all that means the data is already contained in the true.
        throw(ArgumentError("Data already contained in tree, no duplicates allowed."))
    end

    return
end

function remove!(B::BST, data)
    """ This function removes the node containing the specified data.

        throws:
            KeyError: If there is no node containing the data, including if
            the tree is empty.
    """

    if isequal(B.root,nothing) # if the BST is empty raise a value error
        throw(KeyError(data))
    end

    # this is the node to remove; find() will raise a KeyError if
    # there is no node in the BST containing the data.
    nodeToRemove = find(B,data)

    if isequal(nodeToRemove.left,nothing) && isequal(nodeToRemove.right,nothing) # if the nde to remove is a leaf node.
        if isequal(nodeToRemove.prev, nothing) # if it is the root node
            B.root = nothing
        elseif nodeToRemove.prev.left == nodeToRemove # if the node to remove is a left child of the parent node, set the left pointer to nothing
            nodeToRemove.prev.left=nothing
        else # otherwise it is a right child so set the parent's right child to nothing.
            nodeToRemove.prev.right=nothing
        end
    elseif isequal(nodeToRemove.right,nothing) # if the node to remove has only a left child
        if B.root == nodeToRemove # if the node to remove happens to be the root.
            B.root = B.root.left
            B.root.prev = nothing
        elseif nodeToRemove.prev.left == nodeToRemove # if the node to remove is a left child of its parent node.
            prevNode = nodeToRemove.prev
            prevNode.left = nodeToRemove.left
            prevNode.left.prev = prevNode
        else # otherwise it is a right childof the parent node
            prevNode = nodeToRemove.prev
            prevNode.right = nodeToRemove.left
            prevNode.right.prev = prevNode
        end
    elseif isequal(nodeToRemove.left,nothing) # if the node to remove has only a right child.
        if B.root == nodeToRemove # if this node happens to be the root
            B.root = B.root.right
            B.root.prev = nothing
        elseif nodeToRemove.prev.left == nodeToRemove # if the node to remove is a left child of its parent node.
            prevNode = nodeToRemove.prev
            prevNode.left = nodeToRemove.right
            prevNode.left.prev = prevNode
        else # otherwise it is a right child of the parent node
            prevNode = nodeToRemove.prev
            prevNode.right = nodeToRemove.right
            prevNode.right.prev = prevNode
        end
    else # if the node to remove has two children or is the root node.
        if B.root == nodeToRemove # if the node to remove is the root node (and has two children)
            leaf = B.root.left
            while !isequal(B.right,nothing) # go to the farthes right laft node after going one level left
                leaf = leaf.right
            end
            newData = leaf.data
            remove!(B,leaf.data) # remove the leaf
            self.root.data = newData # set the roots new value
        else # if the node to remove is not the root, follow the same strategy as above.
            leaf = nodeToRemove.left
            while !isequal(B.right,nothing)
                leaf = leaf.right
            end
            newData = leaf.Data
            remove!(B, leaf.Data)
            nodeToRemove.data = newData
        end
    end
    B.nodeCount -= 1

    return
end


function draw(B::TreeStructure)
    if isequal(B.root,nothing)
        return
    end

    nodes = [B.root]
    edges = []
    while length(nodes) > 0
        currentNode = pop!(nodes)
        for childNode in [currentNode.left, currentNode.right]
            if !isequal(childNode,nothing)
                append!(edges, [(currentNode.data, childNode.data)])
                # add_edge!(G,currentNode.data, childNode.data)
                append!(nodes, [childNode])
            end
        end
    end

    el = Edge.(edges)

    G = SimpleDiGraph(el)

    graphplot(G,curves=false)

end

end
